crates/bevy_ecs/src/system/mod.rs

```diff
@@ -747,36 +668,118 @@ mod tests {
         }
     }
 
-    #[test]
-    fn system_state_safety_test() {
-        struct A(usize);
-        struct B(usize);
-
-        struct State {
-            state_r: SystemState<Query<'static, 'static, &'static A>>,
-            state_w: SystemState<Query<'static, 'static, &'static mut A>>,
-        }
-
-        impl State {
-            fn get_component<'w>(&mut self, world: &'w mut World, entity: Entity) {
-                let q1 = self.state_r.get(&world);
-                let a1 = q1.get(entity).unwrap();
-
-                let mut q2 = self.state_w.get_mut(world);
-                let a2 = q2.get_mut(entity).unwrap();
-                // this should fail to compile
-                println!("{}", a1.0);
-            }
-
-            fn get_components<'w>(&mut self, world: &'w mut World) {
-                let q1 = self.state_r.get(&world);
-                let a1 = q1.iter().next().unwrap();
-
-                let mut q2 = self.state_w.get_mut(world);
-                let a2 = q2.iter_mut().next().unwrap();
-                // this should fail to compile
-                println!("{}", a1.0);
-            }
-        }
-    }
+    // NOTE: These tests are commented out because they cannot (and should not) compile due to unsafe call patterns.
```

Comment:
It looks like you can write compile-fail tests in doctests: https://stackoverflow.com/questions/55327185/how-do-i-write-a-test-assuring-a-compilation-error

Could we do that instead so this is less brittle? Ideally with some internal-only docs.

Reply:
It also looks like there's an appropriate crate for this: https://github.com/Manishearth/compiletest-rs

Perhaps that would be useful as a dev-dependency instead?

Reply:
Ooh yeah i dig that.

Reply:
I opted for doc tests as they don't require complicating our testing infrastructure.